### DISCUSSION TITLE
Casts for json fields

### DISCUSSION THREAD

**Discussion by projct1** [laravel/framework#53818] [Ideas]
This is not working now:
```php
class Birthday extends Model
{
    protected $casts = [
        'birthday->day_from' => 'integer',
        'birthday->day_to' => 'integer',
        'birthday->kitchen' => 'boolean',
        'birthday->kitchen_percent' => 'integer',
        'birthday->kitchen_percent_equals_age' => 'boolean',
        'birthday->alcohol' => 'boolean',
        'birthday->alcohol_percent' => 'integer',
        'birthday->alcohol_percent_equals_age' => 'boolean',
        'birthday->hookah' => 'boolean',
        'birthday->hookah_percent' => 'integer',
        'birthday->hookah_percent_equals_age' => 'boolean'
    ];
}
```
Will be good got this feature like `$fillable` json fields https://laravel.com/docs/11.x/eloquent#mass-assignment-json-columns

**Comment by ahinkle** [comment]
Tests are passing. Do you have a reproducible result in which they are not? Or at minimum a repository for reproduction? Thanks!

https://github.com/laravel/framework/blob/2218313a53286d9defd8d5e0155d5556d31748bb/tests/Database/DatabaseEloquentIntegrationTest.php#L1412-L1426

**Reply by projct1** [reply]
1) Clone this repo https://github.com/projct1/laravel-json-casts
2) Run `php artisan migrate`
3) Run `php artisan serve`
4) Open link.

User model contains `casts` function and `$fillable` property:
```php
protected $fillable = [
    'name',
    'email',
    'data',
    'password',
];

protected function casts(): array
{
    return [
        'data' => 'json',
        'email_verified_at' => 'datetime',
        'password' => 'hashed',
        'data->custom_value' => 'bool',
    ];
}
```
But after creating model:
```php
Route::get('/', function () {
    return App\Models\User::create([
        'name' => 'Test',
        'email' => 'test@test.com',
        'password' => Hash::make('password'),
        'data' => [
            'custom_value' => 1
        ]
    ]);
});
```
`Data->custom_value` still 1, not bool.

**Reply by ahinkle** [reply]
Thanks! Yeah, I see what you mean now. I looked through the recent releases but didn’t notice anything that might have caused this. Do you know when you first noticed the issue (between specific versions)?

Also, has this ever worked for casts before? It does make sense, but I’m just trying to understand—when you say it’s not working now, did it previously work?

**Reply by projct1** [reply]
> Do you know when you first noticed the issue (between specific versions)?

I can't say for sure, but approximately after version 11.33.2.

> Also, has this ever worked for casts before? It does make sense, but I’m just trying to understand—when you say it’s not working now, did it previously work?

I started specifying specific keys for json fields in `$fillable` and `$casts` quite recently. Up until now I was just using json:
```php
protected $fillable = [
    'data'
];

protected function casts(): array
{
    return [
        'data' => 'json'
    ];
}
```

By the way, here is similar problem, i guess https://github.com/laravel/framework/issues/53830
